docs/tutorials/time-series.md

```diff
@@ -0,0 +1,738 @@
+   records without depending on a physical clock.
+
+2. **Waterline** of a relation with respect to a timestamp column is a timestamp value
+   `W`, such that all future additions or deletions in the relation will have timestamps
+   greater than or equal to `W`.
+
+### The data retention algorithm
+
+Feldera uses the following (simplified) procedure to identify and discard unused
+data:
+
+1. **Compute waterlines of all program relations** through a combination of offline
+   and online dataflow analysis:
+
+   * Given a SQL table with a timestamp column with the current stream time
+     `T` and lateness `L`, waterline `W` for this column is computed as: `W = T - L`.
+
+   * The waterline of a view column is derived from the waterlines
+     of all its inputs based on the semantics of the view
+     query.  In our example, the waterline of the `daily_max.d` column is
+     computed as `W = TIMESTAMP_TRUNC(Wp, DAY)`, where `Wp` is the waterline
+     of the `purchase.ts` column.
+
+     Not all SQL operators produce outputs monotonic in the input timestamp (and
+     hence have waterlines).  In our example, the `daily_max.d` column has a waterline
+     because the `GROUP BY` clause of the query uses a monotonic function
+     (`TIMESTAMP_TRUNC`) over an input column with a waterline.
+
+2. **Compute state retention bounds**. For each view in the program, the query engine
+   determines how much state it needs to maintain in order to evaluate the view incrementally.
+   Some queries, such as simple projections and filters, do not require keeping any state.
+   Others, such as joins and aggregates, may require storing the state of their input
+   relations.  In this case, the query engine determines **how long the state
+   should be retained** based on the semantics of the query and the waterlines
+   computed at the previous step.
+
+   In our example, the engine infers that the `daily_max` view needs to access
+   records in the `purchase` table only until the `purchase.ts` waterline rolls
+   over a date boundary, after which records for the past date can be discarded.
+   More precisrly, it retains `purchase` records where
+   `TIMESTAMP_TRUNC(purchase.ts) >= TIMESTAMP_TRUNC(Wp)`.
+
+3. **Discard old records**. Feldera garbage collects old records that fall below retention
+   bounds in the background.  Since these records are guaranteed to have no effect on
+   new outputs, it is safe to keep them around until a convenient time when they can be
+   removed with low overhead.  In practice, we incorporate garbage collection in the
+   [log-structured merge tree](https://www.feldera.com/blog/feldera-storage)(LSM) maintenance process.
+
+**Not all SQL operators allow discarding old inputs**. Joins, aggregates, and `DISTINCT`'s
+for example may, depending on their exact join condition or the `GROUP BY` clause,
+require storing their entire inputs permanently.  We list operators for
```

Comment:
may require
but if the group key has fields with waterline, you are fine
perhaps you mean global aggregates?